sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala

```diff
@@ -521,7 +499,8 @@ case class RowToColumnarExec(child: SparkPlan) extends RowToColumnarTransition {
  * to/from columnar formatted data.
  */
 case class ApplyColumnarRulesAndInsertTransitions(
-    columnarRules: Seq[ColumnarRule])
+    columnarRules: Seq[ColumnarRule],
+    outputColumnar: Boolean = false)
```

Comment:
This is the new change. Previously we always assume that the final plan should output row data and add `ColumnarToRowExec` if the plan is columnar. Now the caller side can specify if the final plan should output columnar format or not.

Reply:
Add javadoc?

Reply:
Looks like a good idea.

BTW, since `ApplyColumnarRulesAndInsertTransitions` is used a few time, shall we remove the default value?
```
sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:case class ApplyColumnarRulesAndInsertTransitions(
sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:      ApplyColumnarRulesAndInsertTransitions(sparkSession.sessionState.columnarRules),
sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:    ApplyColumnarRulesAndInsertTransitions(context.session.sessionState.columnarRules),
sql/core/src/test/scala/org/apache/spark/sql/execution/ColumnarRulesSuite.scala:    val rules = ApplyColumnarRulesAndInsertTransitions(
sql/core/src/test/scala/org/apache/spark/sql/execution/ColumnarRulesSuite.scala:    val rules = ApplyColumnarRulesAndInsertTransitions(
```